### PR TITLE
[secrets] Allow Secrets in Create Source

### PR DESCRIPTION
**Pull request by cjubb39**
Introduce two new types:
- In sql-land: `SqlValueOrSecret`, an enum which can hold either a `Value` or a `GlobalId`
- In dataflow-land: `StringOrSecret`, an enum which can hold either a `String` or a `GlobalId`

We put a `SecretsReader` in the catalog so that it's accessible as we get to planning sources and sinks.  Currently we inline during planning -- however, there's a follow-up task (#13017) to pass a `SecretsReader` to storaged so it can inline at the last minute.

### Motivation
#12983 
Part of #11504 

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Can use `SECRET`s for Kafka and CSR with options

### CODE REVIEW

File: src/sql/src/plan/statement/ddl.rs
Comment (cjubb39) [inline src/sql/src/plan/statement/ddl.rs:400]: Thoughts? I haven't thought about it deeply but I believe delaying inlining would require passing through a copy of the `SecretsReader` with `KafkaSourceConnector`

File: src/sql/src/plan/statement/ddl.rs
Comment (cjubb39) [inline src/sql/src/plan/statement/ddl.rs:1984]: similar flag here: inlining _could_ be delayed by passing through a `SecretsReader` ref/copy into `KafkaSinkConnectorBuilder` which the sink can use on its own startup -- but it's unclear to me if that small delay is worth the overhead of plumbing through `SecretsReader`

**Review by cjubb39 — COMMENTED** [review]
A couple design notes I'd like feedback on:
- I chose to make a few functions that need to lookup secrets on startup take a `&SecretsReader`.  I could simplify things and just give the `Catalog` its own `Arc<SecretsReader>` / `SecretsReader` but we'd then need to impl `Clone` for `SecretsReaders` (which is doable -- just more work) and that would allow the `Catalog` access longer than it needs.  Because the Catalog itself doesn't really need access, I think this is the right choice -- but wanted to call it out early rather than address after a few iterations
- I need to touch up a few types after rebasing over the PR that will address #12893 -- I think I've notated all of those with inline comments

File: src/sql/src/normalize.rs
Comment (sploiselle) [inline src/sql/src/normalize.rs:113]: There are a bunch of places you're pulling out things from the inner `Value`, e.g. strings--seems like a `TryInto` would be nice

**Review by sploiselle — COMMENTED** [review]
Lots of cosmetic suggestions.

Big question is if we want to mutate the `KafkaSoureConnector` en route to storage, or if "inling" the values that late is no good.

File: src/secrets/src/lib.rs
Comment (benesch) [reply src/secrets/src/lib.rs:90]: `GlobalId` is copy so taking it by value is always the right thing to do!

Comment (cjubb39) [reply src/secrets/src/lib.rs:90]: oh duh of course it is `Copy` -- and thanks, sean, for the suggestion on simplifying.  you both are right that this is a silly thing to complicate for.  I'll change

File: src/sql/src/normalize.rs
Comment (cjubb39) [reply src/sql/src/normalize.rs:127]: hmm -- I'm not sold on implementing `TryInto` because it's not clear that `SqlMaybeValueId` -> `Value` ought to be the default conversion -- rather than like `SqlMaybeValueId` -> `StringOrGlobalId`.

I don't mind dumping the `try_as_value` impl and just use `try_into_value` everywhere if you think having both is too much.

How does that sound?

File: src/sql/src/plan/statement/ddl.rs
Comment (cjubb39) [reply src/sql/src/plan/statement/ddl.rs:400]: hmm I'd like to get away from the model of changing the types within a struct.  If we don't pass through the `SecretsReader`, I'd imagine we'd do one of the following:
- Fill in a `Option<BTreeMap<String,String>>` (this seems bad and is something nikhil got rid of recently)
- Change all the `MaybeStringId::Secret`s to be `Strings` and assert later that everything is the String variant

If we pass the `MaybeStringId`s along further, I'd prefer we do it all the way through with a `SecretsReader` -- now that we've made `SecretsReader` impl `Clone`!

What do you think?

Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:394]: Ah, I think it could be done on `CatalogState` itself, which already has the `SecretsReader`, i.e. `CatalogState::source_description_for`. That is basically as late as humanly possible. The question remains, though, what the types ought to be. I'm not opposed to literalizing them (i.e. taking them all to `String`), but idk if that's too hacky.

Comment (cjubb39) [reply src/sql/src/plan/statement/ddl.rs:394]: I did a little digging here and doing processing at that point would require a lot of manipulation to do the translation in that function / somewhere around there because the mapping is so many layers deep inside of `CatalogItem`.

And then doing the inline even later, in `render_source`, e.g., would require moving the secrets reader into storage also -- which we probably don't want to do.

From our chat earlier, it seems like you're okay with this so I'll leave it as is

File: src/secrets/src/lib.rs
Comment (benesch) [inline src/secrets/src/lib.rs:89]: I'm confused! What does `SecretsReader` being `Clone` have to do with concurrent reads? And why would concurrent reads be a problem? Something about a hypothetical alternative design of this reader in which it stores a file descriptor rather than a path? What's the rationale behind inlining `fs::read`? Protection against some hypothetical adversarial stdlib implementation that closes and reopens repeatedly while reading?

Was there something wrong with the original code or commentary? ISTM like the original code was correct and worked as described—but maybe I'm missing something!

File: src/secrets/src/lib.rs
Comment (benesch) [inline src/secrets/src/lib.rs:107]: Spurious?

File: src/sql/src/catalog.rs
Comment (benesch) [inline src/sql/src/catalog.rs:191]:
```suggestion
    /// Returns the secrets reader associated with the catalog.
```

File: src/sql/src/plan/statement/ddl.rs
Comment (benesch) [reply src/sql/src/plan/statement/ddl.rs:401]: > And then doing the inline even later, in render_source, e.g., would require moving the secrets reader into storage also -- which we probably don't want to do.

No no, this is _exactly_ what we want to do! We should avoid sending secrets over the wire whenever possible, and the tighter the scope is on a secret that comes out of the reader, the better. Pushing the secret reading into storaged is a great way to accomplish this.

Totally fine to leave this to a follow up PR though.

**Review by benesch — COMMENTED** [review]
Woohoo! This is looking very close to merge-able to me. Deferring the actual approval to Sean, but I left two comments inline.

File: src/secrets/src/lib.rs
Comment (cjubb39) [reply src/secrets/src/lib.rs:89]: If `SecretsReader` is `Clone`, then we can have concurrent reads from disk.

But you're right that the issue I describe isn't really related to reads being concurrent -- the comment is really just clarifying _why_ the current implementation of reading from disk is safe.  (You could imagine, e.g., a silly impl that, if it hits an error, it tries to be helpful and reopens the file and continues reading from where it left off or something)

The old code was correct but I inlined the impl because 1) it's short and 2) it makes it clear the same file handle is used the whole time.  But if that just makes things more confusing, I can "out-line" it

File: src/secrets/src/lib.rs
Comment (cjubb39) [reply src/secrets/src/lib.rs:107]: whoops -- was necessary in a past version but not anymore.  will remove here and above

File: src/sql/src/plan/statement/ddl.rs
Comment (cjubb39) [reply src/sql/src/plan/statement/ddl.rs:401]: sure -- will leave a TODO and an issue

File: src/sql/src/plan/statement/ddl.rs
Comment (cjubb39) [reply src/sql/src/plan/statement/ddl.rs:1959]: and also here

**Comment by philip-stoev** [comment]
I would like to test this a bit Monday.

File: test/kafka-ssl/multi.td
Comment (benesch) [inline test/kafka-ssl/multi.td:108]: 🎉

File: test/kafka-ssl/multi.td
Comment (benesch) [inline test/kafka-ssl/multi.td:69]: We'll have to figure out (in another PR) how to pivot these to be the actual contents of the key rather than the filepath, since users won't be able (by design) to upload files in cloud.

File: src/secrets/src/lib.rs
Comment (benesch) [inline src/secrets/src/lib.rs:92]:
```suggestion
        // Use a `File` handle directly to make clear that we are upholding our documented guarantee
        // of opening and reading from the file only once.
```

File: src/secrets/src/lib.rs
Comment (benesch) [reply src/secrets/src/lib.rs:88]: I think my confusion stems mostly from the comment in the middle, which implies that there is something wrong with `fs::read` that inlining its implementation fixes. I left you a rewording suggestion below that avoids bringing `fs::read` into the mix.

**Review by benesch — APPROVED** [review]
Could you pull the changes to coord.rs out into their own PR? They'll require another cycle or two and a verrrrrry finicky test that we want to get right.

**Comment by benesch** [comment]
@philip-stoev can we do the testing asynchronously? I have several PRs I'd like to whip up over the weekend that are blocked on landing this! The sorts of PRs that will fix #12469, #12472, etc.

**Comment by philip-stoev** [comment]
Ok no problem please proceed as you see fit. I will test after the fact.

**Comment by benesch** [comment]
Wonderful! Just awaiting @sploiselle’s review and the comments I left above! @cjubb39 @sploiselle can we get this merged tomorrow?

**Comment by benesch** [comment]
Oh, Chae is OOO tomorrow. In the interest of forward progress I’ll push this over the line tomorrow, then.

File: src/sql/src/normalize.rs
Comment (sploiselle) [inline src/sql/src/normalize.rs:159]: Should this panic?

**Review by sploiselle — APPROVED** [review]
This really came together. Thanks so much for the hard work on this, @cjubb39. @benesch offered to take this over the line and merge, so I'm approving modulo the few fixes he pointed out and will take care of.

**Comment by sploiselle** [comment]
merged in #13030

**Pull request closed without merging**
